### PR TITLE
`azurerm_arc_kubernetes_cluster` - support for `aad_profile`, `azure_hybrid_benefit`, `arc_agent_auto_upgrade_enabled`, `arc_agent_desired_version` and  `kind`  properties

### PR DESCRIPTION
**Pull request by teowa**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

`azurerm_arc_kubernetes_cluster` - support for `aad_profile`, `azure_hybrid_benefit`, `arc_agent_auto_upgrade_enabled`, `arc_agent_desired_version` and  `kind`  properties

azure doc: part of the https://learn.microsoft.com/en-us/azure/aks/hybrid/aks-create-clusters-portal
swagger: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/hybridkubernetes/resource-manager/Microsoft.Kubernetes/stable/2024-01-01/connectedClusters.json
## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
TF_ACC=1 go test -v ./internal/services/arckubernetes -parallel 5 -run TestAccArcKubernetesCluster_provisionedCluster -timeout 2h -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccArcKubernetesCluster_provisionedClusterBasic
=== PAUSE TestAccArcKubernetesCluster_provisionedClusterBasic
=== RUN   TestAccArcKubernetesCluster_provisionedClusterAadProfile
=== PAUSE TestAccArcKubernetesCluster_provisionedClusterAadProfile
=== CONT  TestAccArcKubernetesCluster_provisionedClusterBasic
=== CONT  TestAccArcKubernetesCluster_provisionedClusterAadProfile
--- PASS: TestAccArcKubernetesCluster_provisionedClusterBasic (248.16s)
--- PASS: TestAccArcKubernetesCluster_provisionedClusterAadProfile (394.13s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/arckubernetes 394.151s
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_arc_kubernetes_cluster` - support for `aad_profile`, `azure_hybrid_benefit`, `arc_agent_auto_upgrade_enabled`, `arc_agent_desired_version` and  `kind`  properties [GH-26916]

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Comment by teowa** [comment]
```
--- PASS: TestAccArcKubernetesCluster_provisionedClusterBasic (228.70s)
--- PASS: TestAccArcKubernetesCluster_provisionedClusterComplete (240.45s)
--- PASS: TestAccArcKubernetesCluster_provisionedClusterAadProfile (321.07s)
--- PASS: TestAccArcKubernetesCluster_provisionedClusterUpdate (325.39s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/arckubernetes 325.412s
```

File: internal/services/arckubernetes/arc_kubernetes_cluster_resource.go [hunk 1]
```diff
@@ -203,12 +309,45 @@ func resourceArcKubernetesClusterUpdate(d *pluginsdk.ResourceData, meta interfac
...
+
+		autoUpgradeOption := arckubernetes.AutoUpgradeOptionsEnabled
+		if !d.Get("arc_agent_auto_upgrade_enabled").(bool) {
+			autoUpgradeOption = arckubernetes.AutoUpgradeOptionsDisabled
+		}
+
+		payload.Properties.ArcAgentProfile.AgentAutoUpgrade = pointer.To(autoUpgradeOption)
+	}
+
+	if err := client.ConnectedClusterCreateThenPoll(ctx, *id, *payload); err != nil {
```
Comment (ms-zhenhua) [inline internal/services/arckubernetes/arc_kubernetes_cluster_resource.go:349]: the function name should be `CreateOrUpdateThenPoll` ?

File: website/docs/r/arc_kubernetes_cluster.html.markdown [hunk 2]
```diff
@@ -46,16 +80,40 @@ The following arguments are supported:
...
-
 * `identity` - (Required) An `identity` block as defined below. Changing this forces a new Arc Kubernetes Cluster to be created.
 
 * `location` - (Required) Specifies the Azure Region where the Arc Kubernetes Cluster should exist. Changing this forces a new Arc Kubernetes Cluster to be created.
 
+* `aad_profile` - (Optional) An `aad_profile` block as specified below.
+
+~> **NOTE** `aad_profile` can only be specified if `kind` is `ProvisionedCluster`.
+
+* `arc_agent_auto_upgrade_enabled` - (Optional) Whether the Arc agents will be upgraded automatically to the latest version. Possible values are `true` and `false`. Defaults to `true`.
```
Comment (ms-zhenhua) [inline website/docs/r/arc_kubernetes_cluster.html.markdown:91]:
```suggestion
* `arc_agent_auto_upgrade_enabled` - (Optional) Whether the Arc agents will be upgraded automatically to the latest version. Defaults to `true`.
```

File: website/docs/r/arc_kubernetes_cluster.html.markdown [hunk 3]
```diff
@@ -46,16 +80,40 @@ The following arguments are supported:
...
+
+* `arc_agent_desired_version` - (Optional) The version of the Arc agents to be installed on the cluster.
+
+* `agent_public_key_certificate` - (Optional) Specifies the base64-encoded public certificate used by the agent to do the initial handshake to the backend services in Azure. Changing this forces a new Arc Kubernetes Cluster to be created.
+
+-> **NOTE** `agent_public_key_certificate` must not be specified if `kind` is `ProvisionedCluster`.
+
+* `azure_hybrid_benefit` - (Optional) Indicates whether Azure Hybrid Benefit is opted in. Possible values are `True`, `False` and `NotApplicable`. Defaults to `NotApplicable`.
+
+* `kind` - (Optional) The kind of the Arc Kubernetes Cluster based on host infrastructure. Possible value is `ProvisionedCluster`.
```
Comment (ms-zhenhua) [inline website/docs/r/arc_kubernetes_cluster.html.markdown:101]:
```suggestion
* `kind` - (Optional) The kind of the Arc Kubernetes Cluster based on host infrastructure. The only possible value is `ProvisionedCluster`.
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]

File: internal/services/arckubernetes/arc_kubernetes_cluster_resource.go (hunk 1 above)
Comment (teowa) [reply internal/services/arckubernetes/arc_kubernetes_cluster_resource.go:349]: There is no `CreateOrUpdateThenPoll` method, and the `UpdateThenPoll` method is a patch method, so only the `CreateThenPoll` method is useful here.

File: website/docs/r/arc_kubernetes_cluster.html.markdown (hunk 3 above)
Comment (teowa) [reply website/docs/r/arc_kubernetes_cluster.html.markdown]: fixed

**Comment by teowa** [comment]
Hi @ms-zhenhua , thanks for reviewing this!
I have changed the code, please kindly take another look.

File: website/docs/r/arc_kubernetes_cluster.html.markdown [hunk 4]
```diff
@@ -46,16 +80,40 @@ The following arguments are supported:
...
+
+* `arc_agent_desired_version` - (Optional) The version of the Arc agents to be installed on the cluster.
+
+* `agent_public_key_certificate` - (Optional) Specifies the base64-encoded public certificate used by the agent to do the initial handshake to the backend services in Azure. Changing this forces a new Arc Kubernetes Cluster to be created.
+
+-> **NOTE** `agent_public_key_certificate` must not be specified if `kind` is `ProvisionedCluster`.
+
+* `azure_hybrid_benefit` - (Optional) Indicates whether Azure Hybrid Benefit is opted in. Possible values are `True`, `False` and `NotApplicable`. Defaults to `NotApplicable`.
+
+* `kind` - (Optional) The kind of the Arc Kubernetes Cluster based on host infrastructure. The only possible value is `ProvisionedCluster`.
```
Comment (ms-zhenhua) [inline website/docs/r/arc_kubernetes_cluster.html.markdown:101]:
```suggestion
* `kind` - (Optional) The kind of the Arc Kubernetes Cluster based on host infrastructure. The only possible value is `ProvisionedCluster`. Changing this forces a new Arc Kubernetes Cluster to be created.
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]

File: website/docs/r/arc_kubernetes_cluster.html.markdown (hunk 4 above)
Comment (teowa) [reply website/docs/r/arc_kubernetes_cluster.html.markdown]: fixed

**Comment by teowa** [comment]
Hi @ms-zhenhua, thanks for reviewing!
I have changed my code, please kindly take another look, thanks!

**Review by ms-zhenhua — APPROVED** [review]

**Comment by ms-zhenhua** [comment]
Hi @teowa ,

Thank you for your updates. LGTM~

**Review by stephybun — CHANGES_REQUESTED** [review]

File: internal/services/arckubernetes/arc_kubernetes_cluster_resource.go [hunk 5]
```diff
@@ -56,15 +58,81 @@ func resourceArcKubernetesCluster() *pluginsdk.Resource {
...
 				ValidateFunc: azValidate.Base64EncodedString,
+				// agentPublicKeyCertificate input must be empty for Connected Cluster of Kind: Provisioned Cluster
+				ConflictsWith: []string{"kind"},
 			},
 
 			"identity": commonschema.SystemAssignedIdentityRequiredForceNew(),
 
 			"location": commonschema.Location(),
 
+			"aad_profile": {
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/arckubernetes/arc_kubernetes_cluster_resource.go]: updated

File: internal/services/arckubernetes/arc_kubernetes_cluster_resource.go [hunk 6]
```diff
@@ -56,15 +58,81 @@ func resourceArcKubernetesCluster() *pluginsdk.Resource {
...
+			"azure_hybrid_benefit": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Default:  string(arckubernetes.AzureHybridBenefitNotApplicable),
+				ValidateFunc: validation.StringInSlice([]string{
+					string(arckubernetes.AzureHybridBenefitTrue),
+					string(arckubernetes.AzureHybridBenefitFalse),
+					string(arckubernetes.AzureHybridBenefitNotApplicable),
+				}, false),
+			},
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/arckubernetes/arc_kubernetes_cluster_resource.go]: changed

File: internal/services/arckubernetes/arc_kubernetes_cluster_resource.go [hunk 7]
```diff
@@ -56,15 +58,81 @@ func resourceArcKubernetesCluster() *pluginsdk.Resource {
...
+
+			"kind": {
+				Type:          pluginsdk.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"agent_public_key_certificate"},
+				ValidateFunc: validation.StringInSlice([]string{
+					string(arckubernetes.ConnectedClusterKindProvisionedCluster),
+				}, false),
+			},
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/arckubernetes/arc_kubernetes_cluster_resource.go]: We need to expose this, this feature did not exist when the arc_kubernetes resource was onboarding with API version 2021-10-01 in https://github.com/hashicorp/terraform-provider-azurerm/pull/15401, by default the `kind` is not set. `kind=ProvisionedCluster` is only used for AKS on HCI, there is a bicep example https://learn.microsoft.com/en-us/azure/aks/hybrid/create-clusters-bicep#bicep-template-file-mainbicep

File: internal/services/arckubernetes/arc_kubernetes_cluster_resource.go [hunk 8]
```diff
@@ -203,12 +309,45 @@ func resourceArcKubernetesClusterUpdate(d *pluginsdk.ResourceData, meta interfac
...
 	}
 
-	props := arckubernetes.ConnectedClusterPatch{
-		Tags: tags.Expand(d.Get("tags").(map[string]interface{})),
+	resp, err := client.ConnectedClusterGet(ctx, *id)
+	if err != nil {
+		return fmt.Errorf("retrieving %s: %+v", id, err)
+	}
+
+	payload := resp.Model
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/arckubernetes/arc_kubernetes_cluster_resource.go]: changed

**Review by stephybun — COMMENTED** [review]
@teowa the more I look through these changes the more I think that this should actually be a separate resource for the following reasons:
- Exposing `kind`/`type` isn't something we typically do, there are old resources in the provider that do this but this is a legacy pattern. We've found splitting these into distinct resources per kind or type makes them easier to maintain and provides a better UX with less confusion
- Introducing `kind` on this resource will introduce confusion for users as there are now some arguments that cannot be set for the `provisionedCluster` kind but have to be set when we're creating the default `connectedCluster` or `azure_active_directory` settings are only available and applicable when we're using the `provisionedCluster` kind

I realise this has been sitting for a while and I should have pointed this out earlier, sometimes it takes a few iterations to realise the gravity of a change. Do you think you could add this as a new resource? I'm unsure of the naming here, the documentation isn't giving me any clear hints as to what a `provisionedCluster` is and how it's different from a `connectedCluster`. Could you provide a bit more information here?

**Comment by teowa** [comment]
Hi @stephybun ,

The ProvisionedCluster I aim to support corresponds to Kubernetes clusters created on Azure Stack HCI.

Based on a verified module release by the internal team, Kubernetes on Azure Stack HCI includes two APIs as referenced in [this module](https://github.com/Azure/terraform-azurerm-avm-res-hybridcontainerservice-provisionedclusterinstance/blob/d8b6827252319e66d1d722dba217714d6cf85484/main.tf#L28-L146):

- Microsoft.Kubernetes/connectedClusters
- Microsoft.HybridContainerService/provisionedClusterInstances

The second API, provisionedClusterInstances, is a child resource of the first one and must have the resource name `default`. I have submitted a PR for this API here: [#27143](https://github.com/hashicorp/terraform-provider-azurerm/pull/27143).

To improve the user experience, we could consider adding these properties into PR#27143. What do you think?

**Comment by stephybun** [comment]
Apologies if I've misunderstood your suggestion, but I'm not sure we can add these properties to the resource in #27143 since that resource maps to an entirely different API endpoint. Looking through the links you've provided on the documentation and the module, I think that the resource in #27143 actually needs to be renamed.

To break it down my suggestion for the Azure API to Terraform resource mapping should be:

- Microsoft.Kubernetes/connectedClusters with kind `provisionedCluster` -> `azurerm_arc_kubernetes_provisioned_cluster`
- Microsoft.HybridContainerService/provisionedClusterInstances -> `azurerm_arc_kubernetes_provisioned_cluster_instance`

Does this make sense to you?

**Comment by teowa** [comment]
Thanks @stephybun , this makes sense, I am going to close this one and support these properties in a new resource `azurerm_arc_kubernetes_provisioned_cluster`

**Comment by stephybun** [comment]
Appreciate it @teowa! Feel free to ping me when it's ready for review

**Comment by teowa** [comment]
Hi @stephybun , I have submitted #28216 to supersede this one.

**Pull request closed without merging**